{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Populating the interactive namespace from numpy and matplotlib\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/lib/python3.6/site-packages/IPython/core/magics/pylab.py:160: UserWarning: pylab import has clobbered these variables: ['random']\n",
      "`%matplotlib` prevents importing * from pylab and numpy\n",
      "  \"\\n`%matplotlib` prevents importing * from pylab and numpy\"\n"
     ]
    }
   ],
   "source": [
    "%pylab inline"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This notebook contains code to reproduce the following from Mitra, et al.:\n",
    "\n",
    "**Main Figures:** \n",
    "* Figure 3A: Number of mutations/RR by category in siblings/probands\n",
    "* Figure 3B: Mutation size by category\n",
    "* Figure 3D: RR by AF bin\n",
    "* Figure 4C: RR by AF bin/locus selection\n",
    "* Figure 4D: RR by AF bin/allele selection\n",
    "\n",
    "**Supplementary Tables:** \n",
    "* Supplementary Table 5: excess by phase\n",
    "* Supplementary Table 6: coding mutations (auto+X)\n",
    "* Supplementary Table 7: deleterious mutations\n",
    "* Supplementary Table 8: expansions (auto+X)\n",
    "\n",
    "**Statistics:**\n",
    "* Mean num proband autosomal mutations\n",
    "* Mean num sibling autosomal mutations\n",
    "* Excess in probands vs. siblings (ttest rel, RR)\n",
    "* Excess in probands vs. siblings after adjust for paternal age (ttest rel)\n",
    "* Excess in probands vs. siblings after adjust for paternal age + remove homopolymers(ttest rel)\n",
    "* Mutation excess in male/female chrX\n",
    "* RR by category"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Input mutation results\n",
    "MUTFILE = \"/storage/ileena/ssc-gangstr-denovos/denovos_GW_priors_Jan20/combined/SSC_allphases_011720_denovos_GW_priors_Jan20.final_qc_mutations.tab\"\n",
    "NAIVEMUTFILE = \"/storage/ileena/ssc-gangstr-denovos/denovos_naive_Jan20/SSC_allphases_011720_denovos_naive_Jan20.final_qc_mutations.tab\"\n",
    "\n",
    "#MUTFILEX = \"/storage/ileena/ssc-gangstr-denovos/denovos_GW_priors_Jan20/SSC_allphases_072820_denovos_model_X.denovos.final_qc_mutations.tab\"\n",
    "MUTFILEX = \"/storage/ileena/ssc-gangstr-denovos/denovos_chrX_naive_Aug20/SSC_allphases_081420_chrX_naive_X.denovos.final_mutations.tab\"\n",
    "NAIVEMUTFILEX = \"/storage/ileena/ssc-gangstr-denovos/denovos_naive_Jan20/SSC_allphases_072820_denovos_naive_X.denovos.final_qc_mutations.tab\"\n",
    "\n",
    "# Make editable in Illustrator\n",
    "import matplotlib\n",
    "matplotlib.rcParams['pdf.fonttype'] = 42\n",
    "matplotlib.rcParams['ps.fonttype'] = 42\n",
    "\n",
    "# Imports\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import os\n",
    "import pandas as pd\n",
    "import scipy.stats\n",
    "import statsmodels.api as sm\n",
    "\n",
    "# Color keys\n",
    "colors = ['gray','red','gold','blue','green','purple','brown','salmon','magenta','cyan','darkblue']\n",
    "\n",
    "# Other data paths\n",
    "ANNOTFILE = \"/storage/mgymrek/ssc-denovos/annotations/hg38_gangstr_MasterAnnotation-rmdup.bed\"\n",
    "AFREQFILE = \"/storage/mgymrek/ssc-denovos/allele-freqs-revision/SSC_merged_allele_counts.tab\"\n",
    "HETFILE = \"/storage/mgymrek/ssc-denovos/allele-freqs-revision/SSC_merged_het.tab\"\n",
    "AGEFILE = \"/storage/ileena/denovos4/metadata/ssc_4phases_ages.csv\"\n",
    "SFARIFILE = \"/storage/mgymrek/ssc-denovos/annotations/SFARI-Gene_genes_08-29-2019release_09-11-2019export.csv\"\n",
    "JACKGENES = \"/storage/mgymrek/ssc-denovos/annotations/jack_102genes.txt\"\n",
    "CONSTRAINTFILE=\"/storage/mgymrek/ssc-denovos/mutea/MUTEA_Constraint_hg38_GangSTR.bed\"\n",
    "\n",
    "# Output path info\n",
    "OUTDIR = \"pdfs/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "# Load expansions\n",
    "ndata = pd.read_csv(NAIVEMUTFILE, sep=\"\\t\")\n",
    "ndata = ndata[(ndata[\"posterior\"]==-1) & (ndata[\"mutsize\"]>=5)] # moderately big expansion events\n",
    "\n",
    "# Filter ones that occur too many times\n",
    "expcounts = ndata.groupby([\"chrom\",\"pos\"], as_index=False).agg({\"child\": len})\n",
    "ndata = pd.merge(ndata, expcounts[expcounts[\"child\"]<=3][[\"chrom\",\"pos\"]], on=[\"chrom\",\"pos\"])\n",
    "\n",
    "# Load main mutations data\n",
    "alldata = pd.read_csv(MUTFILE, sep=\"\\t\")\n",
    "usefams = set(alldata[\"family\"])\n",
    "\n",
    "# Concatenate expansions\n",
    "ndata = ndata[ndata.family.isin(usefams)]\n",
    "alldata = pd.concat([alldata, ndata])\n",
    "\n",
    "# Load chrX\n",
    "alldata_x = pd.read_csv(MUTFILEX, sep=\"\\t\")\n",
    "alldata_x = alldata_x[alldata_x.family.isin(usefams)]\n",
    "ndata_x = pd.read_csv(NAIVEMUTFILEX, sep=\"\\t\")\n",
    "ndata_x = ndata_x[(ndata_x[\"posterior\"]==-1) & (ndata_x[\"mutsize\"]>=5)] # moderately big expansion events\n",
    "expcounts = ndata_x.groupby([\"chrom\",\"pos\"], as_index=False).agg({\"child\": len})\n",
    "ndata_x = pd.merge(ndata_x, expcounts[expcounts[\"child\"]<=3][[\"chrom\",\"pos\"]], on=[\"chrom\",\"pos\"])\n",
    "ndata_x = ndata_x[ndata_x.family.isin(usefams)]\n",
    "alldata_x = pd.concat([alldata_x, ndata_x]).drop_duplicates()\n",
    "\n",
    "# Manually remove problematic families. \n",
    "# 14151 proband had 130 mutations\n",
    "# other families have <20 in both proband and sibling\n",
    "rmfams = [14151, 12434, 12281, 13673, 13351, 13355, 13143]\n",
    "alldata = alldata[~alldata.family.isin(rmfams)]\n",
    "alldata_x = alldata_x[~alldata_x.family.isin(rmfams)]\n",
    "\n",
    "# Load annotation info\n",
    "annot = pd.read_csv(ANNOTFILE, sep=\"\\t\")\n",
    "\n",
    "# Set up fields\n",
    "def GetChrom(x):\n",
    "    if \"chr\" in str(x): return int(x[3:])\n",
    "    else: return int(x)\n",
    "alldata[\"chrom\"] = alldata[\"chrom\"].apply(GetChrom)\n",
    "annot = annot[annot[\"chrom\"].apply(lambda x: x not in [\"X\",\"chrX\",\"Y\",\"chrY\"])]\n",
    "annot[\"chrom\"] = annot[\"chrom\"].apply(GetChrom)\n",
    "annot[\"pos\"] = annot[\"start\"]\n",
    "\n",
    "# Merge\n",
    "adata = pd.merge(alldata, annot, on=[\"chrom\",\"pos\"])\n",
    "\n",
    "# Get age info\n",
    "ages = pd.read_csv(AGEFILE)\n",
    "ages[\"family\"] = ages[\"fam_id\"]\n",
    "ages[\"child\"] = ages[\"sample_id\"]\n",
    "ages[\"phenotype\"] = ages[\"ssc_id\"].apply(lambda x: [1,2][\"p1\" in x])\n",
    "ages = ages[~np.isnan(ages[\"father_age_birth\"])]\n",
    "\n",
    "alldata_x = pd.merge(alldata_x, ages[[\"child\",\"sex\"]], how=\"left\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get Allele freq info\n",
    "het = pd.read_csv(HETFILE, sep=\"\\t\")\n",
    "het[\"chrom\"] = het[\"chrom\"].apply(GetChrom)\n",
    "\n",
    "afreq = pd.read_csv(AFREQFILE, sep=\"\\t\")\n",
    "afreq.columns = [\"chrom\",\"pos\",\"newallele\",\"count\"]\n",
    "afreq[\"chrom\"] = afreq[\"chrom\"].apply(GetChrom)\n",
    "afreq = pd.merge(afreq, het[[\"chrom\",\"pos\",\"total\"]], on=[\"chrom\",\"pos\"])\n",
    "afreq[\"freq\"] = afreq.apply(lambda x: float(x[\"count\"])/x[\"total\"], 1)\n",
    "\n",
    "fdata = pd.merge(adata, afreq, on=[\"chrom\",\"pos\",\"newallele\"], how=\"left\")\n",
    "fdata.ix[np.isnan(fdata[\"freq\"]),\"freq\"] = 0\n",
    "fdata.ix[np.isnan(fdata[\"count\"]),\"count\"] = 0\n",
    "fdata = fdata.drop(\"total\", 1)\n",
    "fdata = pd.merge(fdata, het[[\"chrom\",\"pos\",\"het\",\"total\"]], on=[\"chrom\",\"pos\"])\n",
    "\n",
    "finaldata = pd.merge(fdata, ages[[\"family\",\"child\",\"sex\",\"father_age_birth_years\"]], on=[\"family\",\"child\"], how=\"left\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "AFF_COLOR = \"red\"\n",
    "UNAFF_COLOR = \"blue\"\n",
    "\n",
    "def GetCounts(df, pt, families, sex=None):\n",
    "    famcounts = {}\n",
    "    counts = df[df[\"phenotype\"]==pt]\n",
    "    if sex is not None:\n",
    "        counts = counts[counts[\"sex\"]==sex]\n",
    "    for i in range(counts.shape[0]):\n",
    "        fam = counts[\"family\"].values[i]\n",
    "        count = counts[\"nummut\"].values[i]\n",
    "        famcounts[fam] = count\n",
    "    mutcounts = [famcounts.get(family, 0) for family in families]\n",
    "    return mutcounts\n",
    "\n",
    "def TestBurden(usedata):\n",
    "    # Get counts\n",
    "    families = set(usedata[\"family\"])\n",
    "    child_counts = usedata.groupby([\"child\",\"phenotype\",\"family\",\"father_age_birth_years\"], as_index=False).agg({\"posterior\": len})\n",
    "    child_counts.columns = [\"child\",\"phenotype\",\"family\",\"father_age_birth_years\",\"nummut\"]\n",
    "    aff_counts = GetCounts(child_counts, 2, families)\n",
    "    unaff_counts = GetCounts(child_counts, 1, families)\n",
    "    \n",
    "    # Overall\n",
    "    print(\"#### No adjustment ####\")\n",
    "    print(\"Unaff mean=%s\"%np.mean(unaff_counts))\n",
    "    print(\"Aff mean=%s\"%np.mean(aff_counts))\n",
    "    print(\"Mean diff=%s\"%(np.mean(aff_counts)-np.mean(unaff_counts)))\n",
    "    print(scipy.stats.ttest_rel(unaff_counts, aff_counts))\n",
    "    \n",
    "    # Adjust paternal age\n",
    "    print(\"#### Correct for paternal age ####\")\n",
    "    unaff_counts = child_counts[child_counts[\"phenotype\"]==1]\n",
    "    X = sm.add_constant(np.array(unaff_counts[[\"father_age_birth_years\"]]))\n",
    "    Y = np.array(unaff_counts[\"nummut\"])\n",
    "    model = sm.OLS(Y, X).fit()\n",
    "\n",
    "    # Predict and get residuals for all\n",
    "    child_counts[\"pred_nummut\"] = model.predict(sm.add_constant(np.array(child_counts[[\"father_age_birth_years\"]])))\n",
    "    child_counts[\"resid_nummut\"] = child_counts[\"nummut\"]-child_counts[\"pred_nummut\"]\n",
    "\n",
    "    # Compute p-val\n",
    "    print(scipy.stats.ttest_rel(child_counts[child_counts[\"phenotype\"]==1].sort_values(\"family\")[\"resid_nummut\"],\n",
    "            child_counts[child_counts[\"phenotype\"]==2].sort_values(\"family\")[\"resid_nummut\"]))\n",
    "\n",
    "def PlotMutationSize(fdata_list, fdata_list_labels, figprefix=None):\n",
    "    sizes_means_list_unaff = []\n",
    "    sizes_stderrs_list_unaff = []\n",
    "    sizes_means_list_aff = []\n",
    "    sizes_stderrs_list_aff = []\n",
    "    female_sizes_means_list_unaff = []\n",
    "    female_sizes_means_list_aff = []\n",
    "    male_sizes_means_list_unaff = []\n",
    "    male_sizes_means_list_aff = []\n",
    "\n",
    "    for i in range(len(fdata_list)):\n",
    "        usedata = fdata_list[i]\n",
    "        flabel = fdata_list_labels[i]\n",
    "        print(\"###### %s ######\"%flabel)\n",
    "        # Mutation size\n",
    "        usedata[\"absmutsize\"] = usedata[\"mutsize\"].apply(abs)\n",
    "        sizes_unaff = usedata[usedata[\"phenotype\"]==1][\"absmutsize\"]\n",
    "        sizes_aff = usedata[usedata[\"phenotype\"]==2][\"absmutsize\"]\n",
    "        print(scipy.stats.mannwhitneyu(sizes_unaff, sizes_aff,  alternative=\"less\"))\n",
    "        print(\"Unaff n=%s mean=%s\"%(len(sizes_unaff), np.mean(sizes_unaff)))\n",
    "        print(\"Aff n=%s mean=%s\"%(len(sizes_aff), np.mean(sizes_aff)))\n",
    "\n",
    "        sizes_means_list_unaff.append(np.mean(sizes_unaff))\n",
    "        sizes_stderrs_list_unaff.append(np.sqrt(np.var(sizes_unaff)/len(sizes_unaff)))\n",
    "        sizes_means_list_aff.append(np.mean(sizes_aff))\n",
    "        sizes_stderrs_list_aff.append(np.sqrt(np.var(sizes_aff)/len(sizes_aff)))\n",
    "        \n",
    "        female_sizes_means_list_unaff.append(np.mean(usedata[(usedata[\"phenotype\"]==1)&(usedata[\"sex\"]==\"female\")][\"absmutsize\"]))\n",
    "        female_sizes_means_list_aff.append(np.mean(usedata[(usedata[\"phenotype\"]==2)&(usedata[\"sex\"]==\"female\")][\"absmutsize\"]))\n",
    "        male_sizes_means_list_unaff.append(np.mean(usedata[(usedata[\"phenotype\"]==1)&(usedata[\"sex\"]==\"male\")][\"absmutsize\"]))\n",
    "        male_sizes_means_list_aff.append(np.mean(usedata[(usedata[\"phenotype\"]==2)&(usedata[\"sex\"]==\"male\")][\"absmutsize\"]))\n",
    "        \n",
    "    fig = plt.figure()\n",
    "    fig.set_size_inches((5+len(fdata_list), 5))\n",
    "    ax = fig.add_subplot(111)\n",
    "    \n",
    "    # Plot overall\n",
    "    W = 0.2\n",
    "    xvals = range(len(fdata_list))\n",
    "    ax.bar(xvals, sizes_means_list_unaff, yerr=1.96*np.array(sizes_stderrs_list_unaff), color=UNAFF_COLOR, width=W)\n",
    "    ax.bar([item+W for item in xvals], sizes_means_list_aff, yerr=1.96*np.array(sizes_stderrs_list_aff), color=AFF_COLOR, width=W)\n",
    "    \n",
    "    # Plot by sex\n",
    "    ptsize = 100\n",
    "    ax.scatter([item-W for item in xvals], male_sizes_means_list_unaff, marker=\"s\", color=UNAFF_COLOR, s=ptsize, edgecolors=\"black\")\n",
    "    ax.scatter([item-W for item in xvals], female_sizes_means_list_unaff, marker=\"o\", color=UNAFF_COLOR, s=ptsize, edgecolors=\"black\")\n",
    "    ax.scatter([item+2*W for item in xvals], male_sizes_means_list_aff, marker=\"s\", color=AFF_COLOR, s=ptsize, edgecolors=\"black\")\n",
    "    ax.scatter([item+2*W for item in xvals], female_sizes_means_list_aff, marker=\"o\", color=AFF_COLOR, s=ptsize, edgecolors=\"black\")\n",
    "\n",
    "    ax.set_ylabel(\"Mutation size\", size=15)\n",
    "    ax.spines['right'].set_visible(False)\n",
    "    ax.spines['top'].set_visible(False)\n",
    "    ax.yaxis.set_ticks_position('left')\n",
    "    ax.xaxis.set_ticks_position('bottom');\n",
    "    ax.set_xticks([item+W/2 for item in xvals])\n",
    "    ax.set_xticklabels(fdata_list_labels, rotation=90)\n",
    "    if figprefix is not None:\n",
    "        fig.savefig(os.path.join(OUTDIR, figprefix+\"-mutsize.pdf\"))\n",
    "\n",
    "def PlotMutationCounts(full_fdata, fdata_list, fdata_list_labels, plot=True, figprefix=None):\n",
    "    fdata = full_fdata\n",
    "    \n",
    "    # Get list of families to use\n",
    "    families = set(fdata[\"family\"])\n",
    "    aff_female_families = set(fdata[(fdata[\"phenotype\"]==2) & (fdata[\"sex\"]==\"female\")][\"family\"])\n",
    "    unaff_female_families = set(fdata[(fdata[\"phenotype\"]==1) & (fdata[\"sex\"]==\"female\")][\"family\"])\n",
    "    aff_male_families = set(fdata[(fdata[\"phenotype\"]==2) & (fdata[\"sex\"]==\"male\")][\"family\"])\n",
    "    unaff_male_families = set(fdata[(fdata[\"phenotype\"]==1) & (fdata[\"sex\"]==\"male\")][\"family\"])\n",
    "\n",
    "    W = 0.2\n",
    "    ptsize = 100\n",
    "\n",
    "    # each on a different axis\n",
    "    for i in range(len(fdata_list)):\n",
    "        usedata = fdata_list[i]\n",
    "        flabel = fdata_list_labels[i]\n",
    "        print(\"### Comparing counts for %s #####\"%flabel)\n",
    "        child_counts = usedata.groupby([\"child\",\"phenotype\",\"family\",\"father_age_birth_years\"], as_index=False).agg({\"posterior\": len})\n",
    "        child_counts.columns = [\"child\",\"phenotype\",\"family\",\"father_age_birth_years\",\"nummut\"]\n",
    "        child_counts_by_sex = usedata.groupby([\"child\",\"phenotype\",\"family\",\"sex\"], as_index=False).agg({\"posterior\": len})\n",
    "        child_counts_by_sex.columns = [\"child\",\"phenotype\",\"family\",\"sex\",\"nummut\"]\n",
    "\n",
    "        aff_counts = GetCounts(child_counts, 2, families)\n",
    "        unaff_counts = GetCounts(child_counts, 1, families)\n",
    "        aff_male_counts = GetCounts(child_counts_by_sex, 2, aff_male_families, sex=\"male\")\n",
    "        aff_female_counts = GetCounts(child_counts_by_sex, 2, aff_female_families, sex=\"female\")\n",
    "        unaff_male_counts = GetCounts(child_counts_by_sex, 1, unaff_male_families, sex=\"male\")\n",
    "        unaff_female_counts = GetCounts(child_counts_by_sex, 1, unaff_female_families, sex=\"female\")\n",
    "\n",
    "        print(scipy.stats.ttest_rel(unaff_counts, aff_counts))\n",
    "\n",
    "        if not plot: continue\n",
    "        fig = plt.figure()\n",
    "        fig.set_size_inches((4, 6))\n",
    "        ax = fig.add_subplot(111)\n",
    "        ax.bar([0], [np.mean(unaff_counts)], yerr=[1.96*np.sqrt(np.var(unaff_counts)/len(families))], color=UNAFF_COLOR, width=W)\n",
    "        ax.bar([W], [np.mean(aff_counts)], yerr=[1.96*np.sqrt(np.var(aff_counts)/len(families))], color=AFF_COLOR, width=W)\n",
    "\n",
    "        ax.scatter([-1*W], [np.mean(unaff_male_counts)], marker=\"s\", color=UNAFF_COLOR, s=ptsize, edgecolors=\"black\")\n",
    "        ax.scatter([-1*W], [np.mean(unaff_female_counts)], marker=\"o\", color=UNAFF_COLOR, s=ptsize, edgecolors=\"black\")\n",
    "        ax.scatter([2*W], [np.mean(aff_male_counts)], marker=\"s\", color=AFF_COLOR, s=ptsize, edgecolors=\"black\")\n",
    "        ax.scatter([2*W], [np.mean(aff_female_counts)], marker=\"o\", color=AFF_COLOR, s=ptsize, edgecolors=\"black\")\n",
    "        ax.set_xticks([W/2])\n",
    "        ax.set_ylabel(\"# mutations/child\", size=15)\n",
    "        ax.spines['right'].set_visible(False)\n",
    "        ax.spines['top'].set_visible(False)\n",
    "        ax.yaxis.set_ticks_position('left')\n",
    "        ax.xaxis.set_ticks_position('bottom');\n",
    "    \n",
    "        ax.set_title(fdata_list_labels[i])\n",
    "        if figprefix is not None:\n",
    "            fig.savefig(os.path.join(OUTDIR, figprefix+\"-burden-%s.pdf\"%fdata_list_labels[i]))\n",
    "\n",
    "def TestSexBurdenMW(fdata, usedata, sex):\n",
    "    aff_families = set(fdata[(fdata[\"phenotype\"]==2) & (fdata[\"sex\"]==sex)][\"family\"])\n",
    "    unaff_families = set(fdata[(fdata[\"phenotype\"]==1) & (fdata[\"sex\"]==sex)][\"family\"])\n",
    "\n",
    "    child_counts = usedata[usedata[\"sex\"]==sex].groupby([\"child\",\"phenotype\",\"family\"], as_index=False).agg({\"posterior\": len})\n",
    "    child_counts.columns = [\"child\",\"phenotype\",\"family\",\"nummut\"]\n",
    "    aff_counts = GetCounts(child_counts, 2, aff_families)\n",
    "    unaff_counts = GetCounts(child_counts, 1, unaff_families)\n",
    "    print(\"num aff children=%s\"%len(aff_counts))\n",
    "    print(\"num unaff children=%s\"%len(unaff_counts))\n",
    "    print(\"mean aff=%s\"%np.mean(aff_counts))\n",
    "    print(\"mean unaff=%s\"%np.mean(unaff_counts))\n",
    "    print(scipy.stats.mannwhitneyu(aff_counts, unaff_counts, alternative=\"two-sided\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Mutation counts/burden case vs. control"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "print(\"Burden - overall\")\n",
    "TestBurden(finaldata)\n",
    "\n",
    "print(\"\\nBurden - exclude homopolymers\")\n",
    "TestBurden(finaldata[finaldata[\"period_x\"]>1])\n",
    "\n",
    "print(\"\\nBurden - by phase\")\n",
    "for phase in [\"phase1\", \"phase2\", \"phase3_1\", \"phase3_2\"]:\n",
    "    print(phase)\n",
    "    TestBurden(finaldata[finaldata[\"phase\"]==phase])\n",
    "\n",
    "print(\"\\nBurden - chrX male\")\n",
    "TestSexBurdenMW(finaldata, alldata_x, sex=\"male\")\n",
    "\n",
    "print(\"\\nBurden - chrX female\")\n",
    "TestSexBurdenMW(finaldata, alldata_x, sex=\"female\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Fig. 3A Mutation count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "# Mutation count by category\n",
    "fdata_list_annotations = [finaldata]\n",
    "fdata_list_annotations_nohom = [finaldata[finaldata[\"period_x\"]>1]]\n",
    "fdata_labels_list_annotations = [\"all\"]\n",
    "for col in ['FEATURE-UTR5','FEATURE-UTR3', 'FEATURE-FETALBRAINTSS', 'FEATURE-CODING', 'FEATURE-INTRON', 'FEATURE-FETALBRAINENH']:\n",
    "    fdata_list_annotations.append(finaldata[(finaldata[col]>0)])\n",
    "    fdata_list_annotations_nohom.append(finaldata[(finaldata[col]>0) & (finaldata[\"period_x\"]>1)])\n",
    "    fdata_labels_list_annotations.append(col.split(\"-\")[1])\n",
    "PlotMutationCounts(finaldata, fdata_list_annotations, fdata_labels_list_annotations, figprefix=\"Figure3\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Fig. 3B Relative risk by category"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "import random\n",
    "\n",
    "# binomial method\n",
    "def GetRRLowHigh(aff_counts, unaff_counts):\n",
    "    # return [low, high] for 95% CI\n",
    "    num_unaff = len(unaff_counts)\n",
    "    num_aff = len(aff_counts)\n",
    "    total = np.sum(unaff_counts)+np.sum(aff_counts)\n",
    "    perc = np.sum(aff_counts)/total\n",
    "    perc_se = np.sqrt(perc*(1-perc)/total)\n",
    "    perc_low = perc-1.96*perc_se\n",
    "    perc_high = perc+1.96*perc_se\n",
    "    if perc_low < 0: perc_low = 0\n",
    "    if perc_high > 1: perc_high = 1\n",
    "    try:\n",
    "        rr_low = num_unaff*1.0/num_aff*perc_low/(1-perc_low)\n",
    "        rr_high = num_unaff*1.0/num_aff*perc_high/(1-perc_high)\n",
    "        return rr_low, rr_high\n",
    "    except: return np.nan, np.nan\n",
    "\n",
    "\"\"\"\n",
    "def PlotRR(full_fdata, fdata_list, fdata_list_labels, plot=True, maxORy=None, plotsex=False, figsize=None, figprefix=\"Figure3\",\n",
    "          colors=None):\n",
    "    fdata = full_fdata\n",
    "    \n",
    "    # Get list of families to use\n",
    "    families = set(fdata[\"family\"])\n",
    "    aff_female_families = set(fdata[(fdata[\"phenotype\"]==2) & (fdata[\"sex\"]==\"female\")][\"family\"])\n",
    "    unaff_female_families = set(fdata[(fdata[\"phenotype\"]==1) & (fdata[\"sex\"]==\"female\")][\"family\"])\n",
    "    aff_male_families = set(fdata[(fdata[\"phenotype\"]==2) & (fdata[\"sex\"]==\"male\")][\"family\"])\n",
    "    unaff_male_families = set(fdata[(fdata[\"phenotype\"]==1) & (fdata[\"sex\"]==\"male\")][\"family\"])\n",
    "\n",
    "    # Keep track of the percentage of mutations that are cases/ctrls for each category\n",
    "    RR_all = []\n",
    "    RR_female = []\n",
    "    RR_male = []\n",
    "    \n",
    "    RR_low_all = []\n",
    "    RR_low_female = []\n",
    "    RR_low_male = []\n",
    "    RR_high_all = []\n",
    "    RR_high_female = []\n",
    "    RR_high_male = []\n",
    "    \n",
    "    for i in range(len(fdata_list)):\n",
    "        usedata = fdata_list[i]\n",
    "        flabel = fdata_list_labels[i]\n",
    "        \n",
    "        # Get counts of unaff/aff mutations\n",
    "        child_counts = usedata.groupby([\"child\",\"phenotype\",\"family\",\"father_age_birth_years\"], as_index=False).agg({\"posterior\": len})\n",
    "        child_counts.columns = [\"child\",\"phenotype\",\"family\",\"father_age_birth_years\",\"nummut\"]\n",
    "        child_counts_by_sex = usedata.groupby([\"child\",\"phenotype\",\"family\",\"sex\"], as_index=False).agg({\"posterior\": len})\n",
    "        child_counts_by_sex.columns = [\"child\",\"phenotype\",\"family\",\"sex\",\"nummut\"]\n",
    "        aff_counts = GetCounts(child_counts, 2, families)\n",
    "        unaff_counts = GetCounts(child_counts, 1, families)\n",
    "        aff_male_counts = GetCounts(child_counts_by_sex, 2, aff_male_families, sex=\"male\")\n",
    "        aff_female_counts = GetCounts(child_counts_by_sex, 2, aff_female_families, sex=\"female\")\n",
    "        unaff_male_counts = GetCounts(child_counts_by_sex, 1, unaff_male_families, sex=\"male\")\n",
    "        unaff_female_counts = GetCounts(child_counts_by_sex, 1, unaff_female_families, sex=\"female\")\n",
    "\n",
    "        # Get RR for all/male/female\n",
    "        RR_all.append(np.mean(aff_counts)/np.mean(unaff_counts))\n",
    "        RR_female.append(np.mean(aff_female_counts)/np.mean(unaff_female_counts))\n",
    "        RR_male.append(np.mean(aff_male_counts)/np.mean(unaff_male_counts))\n",
    "\n",
    "        low_all, high_all = GetRRLowHigh(aff_counts, unaff_counts)\n",
    "        RR_low_all.append(low_all)\n",
    "        RR_high_all.append(high_all)\n",
    "        low_male, high_male = GetRRLowHigh(aff_male_counts, unaff_male_counts)\n",
    "        RR_low_male.append(low_male)\n",
    "        RR_high_male.append(high_male)\n",
    "        low_female, high_female = GetRRLowHigh(aff_female_counts, unaff_female_counts)\n",
    "        RR_low_female.append(low_female)\n",
    "        RR_high_female.append(high_female)        \n",
    "        print(\"#### %s #####\"%fdata_list_labels[i])\n",
    "        print(\"Num total/male/female=%s/%s/%s\"%(np.sum(aff_counts)+np.sum(unaff_counts),\n",
    "                                                 np.sum(aff_male_counts)+np.sum(unaff_male_counts),\n",
    "                                               np.sum(aff_female_counts)+np.sum(unaff_female_counts)))\n",
    "\n",
    "    ########### Print results #########\n",
    "    for i in range(len(fdata_list)):\n",
    "        print(\"##### %s #####\"%(fdata_list_labels[i]))\n",
    "        print(\"RR - all: %.2f (%.2f-%.2f)\"%(RR_all[i], RR_low_all[i], RR_high_all[i]))\n",
    "        print(\"RR - male: %.2f (%.2f-%.2f)\"%(RR_male[i], RR_low_male[i], RR_high_male[i]))\n",
    "        print(\"RR - female: %.2f (%.2f-%.2f)\"%(RR_female[i], RR_low_female[i], RR_high_female[i]))\n",
    "        \n",
    "    ########### Plot #########\n",
    "    if not plot: return\n",
    "    xvals = range(len(fdata_list))\n",
    "    fig = plt.figure()\n",
    "    if figsize is None:\n",
    "        fig.set_size_inches((5+len(fdata_list), 3))\n",
    "    else:\n",
    "        fig.set_size_inches(figsize)\n",
    "    ax = fig.add_subplot(111)\n",
    "    W2 = 0.2\n",
    "    RE_err_low = np.array(RR_all)-np.array(RR_low_all)\n",
    "    RE_err_high = np.array(RR_high_all)-np.array(RR_all)\n",
    "    ax.errorbar(xvals, RR_all, yerr = [RE_err_low, RE_err_high], color=\"black\", fmt=\"o\")\n",
    "    if plotsex:\n",
    "        RE_err_low_male = np.array(RR_male)-np.array(RR_low_male)\n",
    "        RE_err_high_male = np.array(RR_high_male)-np.array(RR_male)\n",
    "        RE_err_low_female = np.array(RR_female)-np.array(RR_low_female)\n",
    "        RE_err_high_female = np.array(RR_high_female)-np.array(RR_female)\n",
    "        ax.errorbar([item+W2 for item in xvals], RR_male, yerr = [RE_err_low_male, RE_err_high_male], color=\"green\", fmt=\"o\")\n",
    "        ax.errorbar([item+W2*2 for item in xvals], RR_female, yerr = [RE_err_low_female, RE_err_high_female], color=\"purple\", fmt=\"o\")\n",
    "    if maxORy is not None:\n",
    "        ax.set_ylim(bottom=0.5, top=maxORy)\n",
    "    ax.axhline(y=1, linestyle=\"solid\", linewidth=0.5, color=\"black\")\n",
    "    if plotsex:\n",
    "        ax.set_xticks([item+W2 for item in xvals])\n",
    "    else:\n",
    "        ax.set_xticks(xvals)\n",
    "    ax.set_xticklabels(fdata_list_labels, rotation=90)\n",
    "    ax.set_ylabel(\"RR(#case/#ctrl)\", size=15)\n",
    "    ax.spines['right'].set_visible(False)\n",
    "    ax.spines['top'].set_visible(False)\n",
    "    ax.yaxis.set_ticks_position('left')\n",
    "    ax.xaxis.set_ticks_position('bottom');\n",
    "    if figprefix is not None:\n",
    "        fig.savefig(os.path.join(OUTDIR, figprefix+\"-RR.pdf\"))\n",
    "\"\"\"\n",
    "\n",
    "def PlotRR(full_fdata, fdata_list, fdata_list_labels, plot=True, maxORy=None, plotsex=False, figsize=None, figprefix=\"Figure3\",\n",
    "          colors=None):\n",
    "    fdata = full_fdata\n",
    "    \n",
    "    # Get list of families to use\n",
    "    families = set(fdata[\"family\"])\n",
    "    aff_female_families = set(fdata[(fdata[\"phenotype\"]==2) & (fdata[\"sex\"]==\"female\")][\"family\"])\n",
    "    unaff_female_families = set(fdata[(fdata[\"phenotype\"]==1) & (fdata[\"sex\"]==\"female\")][\"family\"])\n",
    "    aff_male_families = set(fdata[(fdata[\"phenotype\"]==2) & (fdata[\"sex\"]==\"male\")][\"family\"])\n",
    "    unaff_male_families = set(fdata[(fdata[\"phenotype\"]==1) & (fdata[\"sex\"]==\"male\")][\"family\"])\n",
    "\n",
    "    # Keep track of the percentage of mutations that are cases/ctrls for each category\n",
    "    RR_all = []\n",
    "    RR_female = []\n",
    "    RR_male = []\n",
    "    \n",
    "    RR_low_all = []\n",
    "    RR_low_female = []\n",
    "    RR_low_male = []\n",
    "    RR_high_all = []\n",
    "    RR_high_female = []\n",
    "    RR_high_male = []\n",
    "    \n",
    "    p_all = []\n",
    "    p_male = []\n",
    "    p_female = []\n",
    "    \n",
    "    for i in range(len(fdata_list)):\n",
    "        usedata = fdata_list[i]\n",
    "        flabel = fdata_list_labels[i]\n",
    "        \n",
    "        # Get counts of unaff/aff mutations\n",
    "        child_counts = usedata.groupby([\"child\",\"phenotype\",\"family\",\"father_age_birth_years\"], as_index=False).agg({\"posterior\": len})\n",
    "        child_counts.columns = [\"child\",\"phenotype\",\"family\",\"father_age_birth_years\",\"nummut\"]\n",
    "        child_counts_by_sex = usedata.groupby([\"child\",\"phenotype\",\"family\",\"sex\"], as_index=False).agg({\"posterior\": len})\n",
    "        child_counts_by_sex.columns = [\"child\",\"phenotype\",\"family\",\"sex\",\"nummut\"]\n",
    "        aff_counts = GetCounts(child_counts, 2, families)\n",
    "        unaff_counts = GetCounts(child_counts, 1, families)\n",
    "        aff_male_counts = GetCounts(child_counts_by_sex, 2, aff_male_families, sex=\"male\")\n",
    "        aff_female_counts = GetCounts(child_counts_by_sex, 2, aff_female_families, sex=\"female\")\n",
    "        unaff_male_counts = GetCounts(child_counts_by_sex, 1, unaff_male_families, sex=\"male\")\n",
    "        unaff_female_counts = GetCounts(child_counts_by_sex, 1, unaff_female_families, sex=\"female\")\n",
    "\n",
    "        # Get RR for all/male/female\n",
    "        RR_all.append(np.mean(aff_counts)/np.mean(unaff_counts))\n",
    "        RR_female.append(np.mean(aff_female_counts)/np.mean(unaff_female_counts))\n",
    "        RR_male.append(np.mean(aff_male_counts)/np.mean(unaff_male_counts))\n",
    "\n",
    "        # Get p-val for all/male/female\n",
    "        p_all.append(scipy.stats.ttest_rel(unaff_counts, aff_counts)[1]/2) # ttest_rel returns 2-sided. convert to 1-sided\n",
    "        p_male.append(scipy.stats.ttest_ind(unaff_male_counts, aff_male_counts)[1]/2) \n",
    "        p_female.append(scipy.stats.ttest_ind(unaff_female_counts, aff_female_counts)[1]/2) \n",
    "\n",
    "        low_all, high_all = GetRRLowHigh(aff_counts, unaff_counts)\n",
    "        RR_low_all.append(low_all)\n",
    "        RR_high_all.append(high_all)\n",
    "        low_male, high_male = GetRRLowHigh(aff_male_counts, unaff_male_counts)\n",
    "        RR_low_male.append(low_male)\n",
    "        RR_high_male.append(high_male)\n",
    "        low_female, high_female = GetRRLowHigh(aff_female_counts, unaff_female_counts)\n",
    "        RR_low_female.append(low_female)\n",
    "        RR_high_female.append(high_female)        \n",
    "        print(\"#### %s #####\"%fdata_list_labels[i])\n",
    "        print(\"Num total/male/female=%s/%s/%s\"%(np.sum(aff_counts)+np.sum(unaff_counts),\n",
    "                                                 np.sum(aff_male_counts)+np.sum(unaff_male_counts),\n",
    "                                               np.sum(aff_female_counts)+np.sum(unaff_female_counts)))\n",
    "\n",
    "    ########### Print results #########\n",
    "    for i in range(len(fdata_list)):\n",
    "        print(\"##### %s #####\"%(fdata_list_labels[i]))\n",
    "        print(\"RR - all: %.2f (%.2f-%.2f) p=%s\"%(RR_all[i], RR_low_all[i], RR_high_all[i], p_all[i]))\n",
    "        print(\"RR - male: %.2f (%.2f-%.2f) p=%s\"%(RR_male[i], RR_low_male[i], RR_high_male[i], p_male[i]))\n",
    "        print(\"RR - female: %.2f (%.2f-%.2f) p=%s\"%(RR_female[i], RR_low_female[i], RR_high_female[i], p_female[i]))\n",
    "        \n",
    "    ########### Plot #########\n",
    "    if not plot: return\n",
    "    xvals = range(len(fdata_list))\n",
    "    fig = plt.figure()\n",
    "    if figsize is None:\n",
    "        fig.set_size_inches((5+len(fdata_list), 3))\n",
    "    else:\n",
    "        fig.set_size_inches(figsize)\n",
    "    ax = fig.add_subplot(111)\n",
    "    W2 = 0.2\n",
    "    RE_err_low = np.array(RR_all)-np.array(RR_low_all)\n",
    "    RE_err_high = np.array(RR_high_all)-np.array(RR_all)\n",
    "    ax.errorbar(xvals, RR_all, yerr = [RE_err_low, RE_err_high], color=\"black\", fmt=\"o\")\n",
    "    if plotsex:\n",
    "        RE_err_low_male = np.array(RR_male)-np.array(RR_low_male)\n",
    "        RE_err_high_male = np.array(RR_high_male)-np.array(RR_male)\n",
    "        RE_err_low_female = np.array(RR_female)-np.array(RR_low_female)\n",
    "        RE_err_high_female = np.array(RR_high_female)-np.array(RR_female)\n",
    "        ax.errorbar([item+W2 for item in xvals], RR_male, yerr = [RE_err_low_male, RE_err_high_male], color=\"green\", fmt=\"o\")\n",
    "        ax.errorbar([item+W2*2 for item in xvals], RR_female, yerr = [RE_err_low_female, RE_err_high_female], color=\"purple\", fmt=\"o\")\n",
    "    if maxORy is not None:\n",
    "        ax.set_ylim(bottom=0.5, top=maxORy)\n",
    "    ax.axhline(y=1, linestyle=\"solid\", linewidth=0.5, color=\"black\")\n",
    "    if plotsex:\n",
    "        ax.set_xticks([item+W2 for item in xvals])\n",
    "    else:\n",
    "        ax.set_xticks(xvals)\n",
    "    ax.set_xticklabels(fdata_list_labels, rotation=90)\n",
    "    ax.set_ylabel(\"RR(#case/#ctrl)\", size=15)\n",
    "    ax.spines['right'].set_visible(False)\n",
    "    ax.spines['top'].set_visible(False)\n",
    "    ax.yaxis.set_ticks_position('left')\n",
    "    ax.xaxis.set_ticks_position('bottom');\n",
    "    if figprefix is not None:\n",
    "        fig.savefig(os.path.join(OUTDIR, figprefix+\"-RR.pdf\"))\n",
    "        \n",
    "PlotRR(finaldata, fdata_list_annotations, fdata_labels_list_annotations, maxORy=4, plotsex=True, figprefix=\"Figure3\")\n",
    "PlotRR(finaldata, fdata_list_annotations_nohom, fdata_labels_list_annotations, maxORy=4, plotsex=True, figprefix=\"ED-Fig8-NoHomopolymers\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Compare if burden different between subdata vs. full data \n",
    "def CompareBurden(fulldata, subdata):\n",
    "    num_total_aff = fulldata[fulldata[\"phenotype\"]==2].shape[0]\n",
    "    num_total_unaff = fulldata[fulldata[\"phenotype\"]==1].shape[0]\n",
    "    num_sub_aff = subdata[subdata[\"phenotype\"]==2].shape[0]\n",
    "    num_sub_unaff = subdata[subdata[\"phenotype\"]==1].shape[0]\n",
    "    # 2x2 table of mutation in annot/not and in proband/not\n",
    "    table = [[num_sub_aff, num_total_aff-num_sub_aff],\n",
    "            [num_sub_unaff, num_total_unaff-num_sub_unaff]]\n",
    "    return scipy.stats.fisher_exact(table, alternative=\"greater\")\n",
    "\n",
    "print(CompareBurden(finaldata, finaldata[finaldata[\"FEATURE-FETALBRAINTSS\"]>0]))\n",
    "print(CompareBurden(finaldata[finaldata[\"sex\"]==\"male\"], finaldata[(finaldata[\"sex\"]==\"male\") & (finaldata[\"FEATURE-FETALBRAINTSS\"]>0)]))\n",
    "print(CompareBurden(finaldata[finaldata[\"sex\"]==\"female\"], finaldata[(finaldata[\"sex\"]==\"female\") & (finaldata[\"FEATURE-FETALBRAINTSS\"]>0)]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Fig.  3B Mutation size"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Mutation size by category\n",
    "PlotMutationSize(fdata_list_annotations, fdata_labels_list_annotations, figprefix=\"Figure3\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Figure 4A: Relative risk by AF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "mintotal = 0.8*max(fdata[\"total\"])\n",
    "\n",
    "# All\n",
    "fdata_list = [finaldata]\n",
    "fdata_list_nohom = [finaldata[finaldata[\"period_x\"]>1]]\n",
    "fdata_labels_list = [\"All\"]\n",
    "\n",
    "# Other freq bins\n",
    "for freqpair in [(0.05, 1.1), (0.01, 0.05), (0.001, 0.01), (-1,0.001), (-1, 0)]:\n",
    "    minfreq, maxfreq = freqpair\n",
    "    if minfreq == -1 and maxfreq > 0:\n",
    "        fdata_list.append(finaldata[(finaldata[\"total\"]>=mintotal) & (finaldata[\"freq\"]>0) & (finaldata[\"freq\"]<=maxfreq)])\n",
    "        fdata_list_nohom.append(finaldata[(finaldata[\"period_x\"]>1) & (finaldata[\"total\"]>=mintotal) & (finaldata[\"freq\"]>0) & (finaldata[\"freq\"]<=maxfreq)])\n",
    "        fdata_labels_list.append(\"Count1-%s\"%(maxfreq))\n",
    "    else:\n",
    "        fdata_list.append(finaldata[(finaldata[\"total\"]>=mintotal) & (finaldata[\"freq\"]>minfreq) & (finaldata[\"freq\"]<=maxfreq)])\n",
    "        fdata_list_nohom.append(finaldata[(finaldata[\"period_x\"]>1) &(finaldata[\"total\"]>=mintotal) & (finaldata[\"freq\"]>minfreq) & (finaldata[\"freq\"]<=maxfreq)])\n",
    "        fdata_labels_list.append(\"%s-%s\"%(minfreq,maxfreq))\n",
    "        \n",
    "PlotRR(finaldata, fdata_list, fdata_labels_list, plotsex=False, figprefix=\"Figure4-AF\")\n",
    "PlotMutationCounts(finaldata, fdata_list, fdata_labels_list, plot=False, figprefix=None)\n",
    "\n",
    "PlotRR(finaldata, fdata_list_nohom, fdata_labels_list, plotsex=True, figprefix=\"Figure4-AF-nohom\")\n",
    "PlotMutationCounts(finaldata, fdata_list_nohom, fdata_labels_list, plot=False, figprefix=None)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(CompareBurden(finaldata, finaldata[(finaldata[\"freq\"]==0) & (finaldata[\"total\"]>=mintotal)]))\n",
    "print(CompareBurden(finaldata[finaldata[\"sex\"]==\"male\"], finaldata[(finaldata[\"sex\"]==\"male\") & (finaldata[\"freq\"]==0) & (finaldata[\"total\"]>=mintotal)]))\n",
    "print(CompareBurden(finaldata[finaldata[\"sex\"]==\"female\"], finaldata[(finaldata[\"sex\"]==\"female\") & (finaldata[\"freq\"]==0) & (finaldata[\"total\"]>=mintotal)]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Figure 4b-c: RR by AF bin/locus selection"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "import statsmodels.stats.multitest\n",
    "\n",
    "mintotal = 0.8*max(fdata[\"total\"])\n",
    "\n",
    "# Load selection estimates and generate per-allele scores\n",
    "def GetSelectionScore(x):\n",
    "    if np.isnan(x[\"ABC_s_median\"]): return np.nan\n",
    "    diff = abs(x[\"newallele\"]-x[\"optimal_ru\"])\n",
    "    s_est = x[\"ABC_s_median\"]\n",
    "    return diff*s_est\n",
    "\n",
    "#seldata = pd.read_csv(\"/storage/BonnieH/selection_project/per_locus/results/per_locus_results_prelim_all_per.txt\", sep=\"\\t\")\n",
    "SSCFILE = \"/storage/BonnieH/selection_project/per_locus/SISTR_results/0.005_3_0.3_yny5_eurodem_prior2_all_per_0810.txt\"\n",
    "seldata = pd.read_csv(SSCFILE, sep=\"\\t\")\n",
    "seldata[\"pos\"] = seldata[\"start\"]\n",
    "seldata = seldata[~np.isnan(seldata[\"LRT_p_value\"])]\n",
    "print(\"Found SISTR data for %s STRs\"%seldata.shape[0])\n",
    "seldata[\"ABC_upper\"] = seldata[\"ABC_s_95%_CI\"].apply(lambda x: float(x.split(\",\")[1].split(\")\")[0]))\n",
    "seldata[\"ABC_lower\"] = seldata[\"ABC_s_95%_CI\"].apply(lambda x: float(x.split(\",\")[0].split(\"(\")[1]))\n",
    "seldata[\"width\"] = seldata[\"ABC_upper\"]-seldata[\"ABC_lower\"]\n",
    "seldata = seldata[seldata.apply(lambda x:  (x[\"width\"]<0.3), 1)] # scores not trustworthy if width is big\n",
    "\n",
    "print(\"Found SISTR data for %s STRs after filtering\"%seldata.shape[0])\n",
    "\n",
    "rej, qvals = statsmodels.stats.multitest.fdrcorrection(seldata[\"LRT_p_value\"])\n",
    "seldata[\"qvalue\"] = qvals\n",
    "\n",
    "sistr = seldata.copy() # for use downstream, before we merge\n",
    "\n",
    "# Now merge with mutation data\n",
    "seldata = pd.merge(finaldata, seldata, on=[\"chrom\",\"pos\"], how=\"left\")\n",
    "seldata[\"pscore\"] = seldata.apply(GetSelectionScore, 1)\n",
    "seldata[\"total\"] = seldata[\"total_x\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "# All\n",
    "q_thresh = 0.01\n",
    "\n",
    "fdata_list = [seldata, seldata[(seldata[\"qvalue\"]<q_thresh) & ~np.isnan(seldata[\"LRT_p_value\"])] ]\n",
    "fdata_labels_list = [\"All\", \"All-selp\"]\n",
    "colors = [\"black\",\"blue\"]\n",
    "\n",
    "# Other freq bins\n",
    "for freqpair in [(0.05, 1.1), (0.01, 0.05), (0.001, 0.01), (-1,0.001), (-1, 0)]:\n",
    "    minfreq, maxfreq = freqpair\n",
    "    if minfreq == -1 and maxfreq > 0:\n",
    "        fdata_list.append(seldata[(seldata[\"total\"]>=mintotal) & (seldata[\"freq\"]>0) & (seldata[\"freq\"]<=maxfreq)])\n",
    "        fdata_labels_list.append(\"Count1-%s\"%(maxfreq))\n",
    "        colors.append(\"black\")\n",
    "        \n",
    "        fdata_list.append(seldata[(seldata[\"qvalue\"]<q_thresh) & ~np.isnan(seldata[\"LRT_p_value\"]) &\n",
    "                                  (seldata[\"total\"]>=mintotal) & (seldata[\"freq\"]>0) & (seldata[\"freq\"]<=maxfreq)])\n",
    "        fdata_labels_list.append(\"Count1-%s-selp\"%(maxfreq))\n",
    "        colors.append(\"blue\")\n",
    "\n",
    "    else:\n",
    "        fdata_list.append(seldata[(seldata[\"total\"]>=mintotal) & (seldata[\"freq\"]>minfreq) & (seldata[\"freq\"]<=maxfreq)])\n",
    "        fdata_labels_list.append(\"%s-%s\"%(minfreq,maxfreq))\n",
    "        colors.append(\"black\")\n",
    "        \n",
    "        fdata_list.append(seldata[(seldata[\"qvalue\"]<q_thresh) & \n",
    "                (seldata[\"total\"]>=mintotal) & (seldata[\"freq\"]>minfreq) & (seldata[\"freq\"]<=maxfreq)])\n",
    "        fdata_labels_list.append(\"%s-%s-selp\"%(minfreq,maxfreq))\n",
    "        colors.append(\"blue\")\n",
    "        \n",
    "PlotRR(finaldata, fdata_list, fdata_labels_list, plotsex=False, maxORy=2, figprefix=\"Figure4-LocusSel\", figsize=(6, 4))\n",
    "\n",
    "# Per allele score, by af\n",
    "#s_threshs = [0, 10e-5, 10e-4, 5e-3, 0.015]\n",
    "s_threshs = np.percentile(seldata[~np.isnan(seldata[\"pscore\"])][\"pscore\"], [0, 90, 99])\n",
    "for freqbin in [(-0.5,1.1), (0, 0.01), (-1, 0)]:\n",
    "    print(\"\\n\\n%s\"%str(freqbin))\n",
    "    fdata_list = []\n",
    "    fdata_list_labels = []\n",
    "    minfreq, maxfreq = freqbin\n",
    "    for sthresh in s_threshs:\n",
    "        if sthresh == 0: use_q = 1\n",
    "        else: use_q = q_thresh\n",
    "        fdata_list.append(seldata[(seldata[\"freq\"]>minfreq) & (seldata[\"freq\"]<=maxfreq) &\n",
    "                                  (seldata[\"qvalue\"]<use_q) & \n",
    "                                  (seldata[\"pscore\"]>=sthresh) & (seldata[\"total\"]>=mintotal)])\n",
    "        fdata_list_labels.append(\"%.6f\"%sthresh)\n",
    "    print(fdata_list_labels)\n",
    "    PlotRR(finaldata, fdata_list, fdata_list_labels, plotsex=False,\n",
    "           figprefix=\"Figure4-AlleleSel-%s\"%str(freqbin), figsize=(6, 4))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Figure 4b - MUTEA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "MUTEAFILE = \"/storage/mgymrek/ssc-denovos/mutea/MUTEA_Constraint_hg38_GangSTR.bed\"\n",
    "mutea = pd.read_csv(MUTEAFILE, sep=\"\\t\", names=[\"chrom\",\"pos\",\"end\",\"MUTEA\"])\n",
    "mutea[\"chrom\"] = mutea[\"chrom\"].apply(GetChrom)\n",
    "mdata = pd.merge(sistr, mutea, on=[\"chrom\",\"pos\"])\n",
    "\n",
    "mdata = mdata[~np.isnan(mdata[\"ABC_s_median\"])]\n",
    "plotdata = []\n",
    "\n",
    "plotdata.append(mdata[mdata[\"qvalue\"]>=0.01][\"MUTEA\"])\n",
    "plotdata.append(mdata[mdata[\"qvalue\"]<0.01][\"MUTEA\"])\n",
    "\n",
    "fig = plt.figure()\n",
    "ax = fig.add_subplot(111)\n",
    "bp = ax.boxplot(plotdata, sym=\"\", positions=range(len(plotdata)), patch_artist=True,\n",
    "               medianprops={\"color\": \"white\"});\n",
    "for patch in bp[\"boxes\"]:\n",
    "    patch.set_facecolor(\"black\")\n",
    "ax.axhline(y=0, color=\"black\", linewidth=0.5)\n",
    "ax.spines['right'].set_visible(False)\n",
    "ax.spines['top'].set_visible(False)\n",
    "ax.yaxis.set_ticks_position('left')\n",
    "ax.xaxis.set_ticks_position('bottom')\n",
    "ax.set_ylabel(\"MUTEA Z-score\")\n",
    "ax.set_xticks(range(len(plotdata)))\n",
    "ax.set_xticklabels([\"SISTR q>=0.01\", \"SISTR q<0.01\"], rotation=45);\n",
    "fig.savefig(\"pdfs/SISTR_vs_MUTEA.pdf\")\n",
    "\n",
    "import scipy.stats\n",
    "print(scipy.stats.mannwhitneyu(plotdata[0], plotdata[1], alternative=\"greater\"))\n",
    "print(len(plotdata[0]))\n",
    "print(len(plotdata[1]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# ED Fig. 10d-e"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gnomad = pd.read_csv(\"/storage/mgymrek/workspace/sistr-analyses/annotations/gnomad.v2.1.1.lof_metrics.by_gene.txt\", sep=\"\\t\")\n",
    "gnomad[\"gene-coding\"] = gnomad[\"gene\"]\n",
    "\n",
    "annot = pd.read_csv(\"/storage/mgymrek/ssc-denovos/annotations/hg38_gangstr_MasterAnnotation-rmdup.bed\", sep=\"\\t\")\n",
    "annot[\"chrom\"] = annot[\"chrom\"].apply(GetChrom)\n",
    "annot[\"start\"] = annot[\"pos\"]\n",
    "usedata = pd.merge(sistr, annot, on=[\"chrom\",\"start\"], how=\"left\")\n",
    "\n",
    "pli = pd.merge(usedata, gnomad[[\"gene-coding\",\"pLI\",\"oe_lof_upper\",\"mis_z\"]], on=[\"gene-coding\"])\n",
    "pli[\"color\"] = pli.apply(lambda x: [\"black\",\"red\"][int(x[\"LRT_p_value\"]<0.01)], 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for feature in [\"FEATURE-CODING\"]:\n",
    "    pos = pli[(pli[\"qvalue\"]<0.01) & ( (pli[feature]>0)) & ~np.isnan(pli[\"oe_lof_upper\"])]\n",
    "\n",
    "    neg = pli[(pli[\"qvalue\"]>=0.01) & ((pli[feature]>0)) & ~np.isnan(pli[\"oe_lof_upper\"])]\n",
    "\n",
    "    print(pos.shape[0])\n",
    "    print(neg.shape[0])\n",
    "    al = 0.5\n",
    "    fig = plt.figure()\n",
    "    ax = fig.add_subplot(111)\n",
    "    bins = np.arange(-6, 6, 0.25)\n",
    "    ax.hist(pos[\"mis_z\"], color=\"red\", alpha=al, bins=bins, normed=True);\n",
    "    ax.hist(neg[\"mis_z\"], color=\"black\", alpha=al, bins=bins, normed=True);\n",
    "    ax.axvline(np.median(pos[\"mis_z\"]), color=\"red\")\n",
    "    ax.axvline(np.median(neg[\"mis_z\"]), color=\"black\")\n",
    "    ax.spines['right'].set_visible(False)\n",
    "    ax.spines['top'].set_visible(False)\n",
    "    ax.yaxis.set_ticks_position('left')\n",
    "    ax.xaxis.set_ticks_position('bottom')\n",
    "    fig.savefig(\"pdfs/EDFig10_misZ-%s.pdf\"%feature)\n",
    "    print(scipy.stats.mannwhitneyu(pos[\"mis_z\"], neg[\"mis_z\"], alternative=\"greater\"))\n",
    "\n",
    "    fig = plt.figure()\n",
    "    ax = fig.add_subplot(111)\n",
    "    bins = np.arange(0, 1.11, 0.1)\n",
    "    ax.hist(pos[\"pLI\"], color=\"red\", alpha=al, bins=bins, normed=True);\n",
    "    #ax.hist(med[\"pLI\"], color=\"blue\", alpha=al, bins=bins, normed=True);\n",
    "    ax.hist(neg[\"pLI\"], color=\"black\", alpha=al, bins=bins, normed=True);\n",
    "    ax.axvline(np.median(pos[\"pLI\"]), color=\"red\")\n",
    "    #ax.axvline(np.median(med[\"pLI\"]), color=\"blue\")\n",
    "    ax.axvline(np.median(neg[\"pLI\"]), color=\"black\")\n",
    "    ax.spines['right'].set_visible(False)\n",
    "    ax.spines['top'].set_visible(False)\n",
    "    ax.yaxis.set_ticks_position('left')\n",
    "    ax.xaxis.set_ticks_position('bottom')\n",
    "    fig.savefig(\"pdfs/EDFig10_pli-%s.pdf\"%feature)\n",
    "    print(scipy.stats.mannwhitneyu(pos[\"pLI\"], neg[\"pLI\"], alternative=\"greater\"))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Supplementary Table 7 - TRs under selection"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cols = [\"phenotype\",\"chrom\",\"pos\",\"mutsize\",\"motif_x\",\"pscore\",\"TSS_GENE\"]\n",
    "seldata[(seldata[\"pscore\"]>s_threshs[-1]) & (seldata[\"qvalue\"]<q_thresh) & (seldata[\"freq\"]<=0)].sort_values([\"phenotype\",\"pscore\"], ascending=False)[cols]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Supplementary Table 6: Coding mutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "finaldata[finaldata[\"FEATURE-CODING\"]>0] # autosome only. add chrX separately. should be 32 autosomal coding"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Supplementary Table 8 - expansions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ndata # add chrX separately."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
